[export] Enforce ordering for graph inputs of ExportedProgram

yiming0416 opened the pull request:
Summary:
As title. The graph inputs of have the following order:

token -> parameter -> buffer (persistent) -> buffer (non-persistent) -> tensor_constant -> custom_obj -> user_inputs

Verifier is also updated to check this order.

Test Plan: buck2 run @mode/dev-nosan caffe2/test:test_export -- -r test_enforcing_placeholder_order

Differential Revision: D69858068

zhxchen17 commented:
@angelayi any case this will be BC breaking? I couldn't think of any

angelayi commented on torch/_export/utils.py:
```diff
@@ -1350,3 +1350,69 @@ def default_flatten_fn_spec(obj, spec) -> list[Any]:
     _deregister_pytree_flatten_spec(cls)
+
+
+def _enforce_placeholder_order_pass(
```
can you put this in torch/_export/passes?

angelayi commented on torch/_export/utils.py:
```diff
@@ -1350,3 +1350,69 @@ def default_flatten_fn_spec(obj, spec) -> list[Any]:
+    with gm.graph.inserting_before(input_nodes[0]):
+        new_placeholder_nodes = []
+        for input_spec, input_node in zip(reordered_input_specs, reordered_input_nodes):
+            new_node = gm.graph.placeholder(input_node.name)
+            # overwrite new_node.name to be the same as input_node.name
+            # in case there is naming collision and suffix is added
+            new_node.name = input_node.name
+            new_node.meta = input_node.meta
+            new_placeholder_nodes.append(new_node)
```
Could you just do:
```python
new_node = gm.graph.node_copy(input_node)
input_node.replace_all_uses_with(new_node)
gm.graph.erase_node(input_node)
new_node.name = input_node.name
```

angelayi commented on torch/_export/utils.py:
```diff
@@ -1350,3 +1350,69 @@ def default_flatten_fn_spec(obj, spec) -> list[Any]:
+    assert len(reordered_input_specs) == len(graph_signature.input_specs)
+    assert len(reordered_input_nodes) == len(input_nodes)
+
+    with gm.graph.inserting_before(input_nodes[0]):
```
I believe just doing `gm.graph.inserting_before()` will put the nodes at the very top

angelayi commented on torch/_export/verifier.py:
```diff
@@ -420,6 +420,35 @@ def _verify_exported_program_signature(exported_program) -> None:
+    for input_spec in gs.input_specs:
+        kind_index = expected_order.index(input_spec.kind)
+        if kind_index < current_order_index:
+            raise SpecViolationError(
+                f"InputSpec kind {input_spec.kind} is out of order. Expected order is {expected_order}."
```
I believe this is BC breaking if you add the verifier check for all exported programs, because saved exported programs who do not have this ordering will fail that check when loading.

yiming0416 committed:
[export] Enforce ordering for graph inputs of ExportedProgram (#147505)

Summary:

As title. The graph inputs of have the following order:

token -> parameter -> buffer (persistent) -> buffer (non-persistent) -> tensor_constant -> custom_obj -> user_inputs

Verifier is also updated to check this order.

buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_enforcing_placeholder_order

Test Plan:
```
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_enforcing_placeholder_order
```

Differential Revision: D69858068

angelayi commented on test/export/test_export.py:
```diff
@@ -10515,6 +10515,47 @@ def forward(self, x, y, div="floor"):
         self.assertEqual(div_spec.arg.name, "div")
         self.assertEqual(div_spec.arg.value, "floor")
 
+    def test_enforcing_placeholder_order(self):
```
youre not calling enforce_placeholder_order_pass :P
